shared/chat/conversation/input-area/mention-handler-hoc/index.desktop.js

```diff
@@ -0,0 +1,251 @@
+
+type MentionHocState = {
+  upArrowCounter: number,
+  downArrowCounter: number,
+  pickSelectedCounter: number,
+  mentionFilter: string,
+  channelMentionFilter: string,
+  mentionPopupOpen: boolean,
+  channelMentionPopupOpen: boolean,
+}
+
+const mentionHoc = (InputComponent: React.ComponentType<Props>) => {
+  class MentionHoc extends React.Component<PropsFromContainer, MentionHocState> {
+    state: MentionHocState = {
+      upArrowCounter: 0,
+      downArrowCounter: 0,
+      pickSelectedCounter: 0,
+      mentionFilter: '',
+      channelMentionFilter: '',
+      mentionPopupOpen: false,
+      channelMentionPopupOpen: false,
+    }
+    _inputRef: ?any
+
+    _setMentionPopupOpen = (mentionPopupOpen: boolean) => {
+      this.setState({mentionPopupOpen})
+    }
+
+    _setChannelMentionPopupOpen = (channelMentionPopupOpen: boolean) => {
+      this.setState({channelMentionPopupOpen})
+    }
+
+    _setMentionFilter = (mentionFilter: string) => {
+      this.setState({mentionFilter})
+    }
+
+    _setChannelMentionFilter = (channelMentionFilter: string) => {
+      this.setState({channelMentionFilter})
+    }
+
+    inputSetRef = (input: any) => {
+      this.props._inputSetRef(input)
+      this._inputRef = input
+    }
+
+    insertMention = (u: string, options?: {notUser: boolean}) => {
+      this._replaceWordAtCursor(`@${u} `)
+      this._setMentionPopupOpen(false)
+
+      // This happens if you type @notausername<enter>. We've essentially 'picked' nothing and really want to submit
+      // This is a little wonky cause this component doesn't directly know if the list is filtered all the way out
+      if (options && options.notUser) {
+        if (this.props.text) {
+          this._onSubmit()
+        }
+      }
+    }
+
+    switchMention = (u: string) => {
+      this._replaceWordAtCursor(`@${u}`)
+    }
+
+    insertChannelMention = (c: string, options?: {notChannel: boolean}) => {
+      this._replaceWordAtCursor(`#${c} `)
+      this._setChannelMentionPopupOpen(false)
+
+      // This happens if you type #notachannel<enter>. We've essentially 'picked' nothing and really want to submit
+      // This is a little wonky cause this component doesn't directly know if the list is filtered all the way out
+      if (options && options.notChannel) {
+        if (this.props.text) {
+          this._onSubmit()
+        }
+      }
+    }
+
+    switchChannelMention = (c: string) => {
+      this._replaceWordAtCursor(`#${c}`)
+    }
+
+    _triggerUpArrowCounter = () => {
+      this.setState(({upArrowCounter}) => ({upArrowCounter: upArrowCounter + 1}))
+    }
+
+    _triggerDownArrowCounter = () => {
+      this.setState(({downArrowCounter}) => ({downArrowCounter: downArrowCounter + 1}))
+    }
+
+    _triggerPickSelectedCounter = () => {
+      this.setState(({pickSelectedCounter}) => ({pickSelectedCounter: pickSelectedCounter + 1}))
+    }
+
+    onKeyDown = (e: SyntheticKeyboardEvent<>) => {
+      this.props._onKeyDown(e)
```

Comment:
🍳 looks like up/down/tab isn't working to change selection

Reply:
fixed here in part 10: https://github.com/keybase/client/pull/10755/files#diff-e1e43dcc902e8130f21aadad81183422L148
due to the withState -> withStateHandlers update